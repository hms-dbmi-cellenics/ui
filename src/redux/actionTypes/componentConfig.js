/**
 * Sets plot configs and data.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const RESET_CONFIG = `${COMPONENT_CONFIG}/reset`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;
const PLOT_DATA_LOADED = `${COMPONENT_CONFIG}/plotDataLoaded`;
const PLOT_DATA_LOADING = `${COMPONENT_CONFIG}/plotDataLoading`;
const PLOT_DATA_ERROR = `${COMPONENT_CONFIG}/plotDataError`;

export {
  LOAD_CONFIG, UPDATE_CONFIG, RESET_CONFIG, SAVE_CONFIG,
  PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR,
};
