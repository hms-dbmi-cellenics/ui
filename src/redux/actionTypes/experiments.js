const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

const EXPERIMENTS_SWITCH = `${EXPERIMENTS}/switch`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SWITCH,
};
