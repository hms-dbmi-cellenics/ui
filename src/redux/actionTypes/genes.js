const GENES = 'genes';

/**
 * Turns on the loading condition for gene information.
 * Components that have the particular properties requested in their dependencies
 * should set a loading state at this event.
 */
const GENES_PROPERTIES_LOADING = `${GENES}/propertiesLoading`;

/**
 * Sets the state of the gene data to be successfully loaded, with the appropriate gene data.
 * This is meant to be used for paginated data, where the data is
 * lazy-loaded by an API call or equivalent.
 */
const GENES_PROPERTIES_LOADED_PAGINATED = `${GENES}/propertiesLoadedPaginated`;

/**
 * Sets an error condition for gene properties data.
 */
const GENES_PROPERTIES_ERROR = `${GENES}/propertiesError`;

/**
 * Updates the list of selected genes with a new selected gene.
 */
const GENES_SELECT = `${GENES}/select`;

/**
 * Removes a deselected gene from the list of selected genes.
 */
const GENES_DESELECT = `${GENES}/deselect`;

/**
 * Turns on the loading state for expression data.
 */
const GENES_EXPRESSION_LOADING = `${GENES}/expressionLoading`;

/**
 * Sets the state of the expression store to be successfully loaded,
 * with the appropriate expression data.
 */
const GENES_EXPRESSION_LOADED = `${GENES}/expressionLoaded`;

/**
 * Sets an error condition for gene expression data.
 */
const GENES_EXPRESSION_ERROR = `${GENES}/expressionError`;

/**
 * Sets marker genes as loading
 */
const MARKER_GENES_LOADING = `${GENES}/markerGenesLoading`;

/**
 * Marker genes have finished loading
 */
const MARKER_GENES_LOADED = `${GENES}/markerGenesLoaded`;

/**
 * Marker genes load task had some kind of error
 */
const MARKER_GENES_ERROR = `${GENES}/markerGenesError`;

/**
 * Downsampled genes expression load task
 */
const DOWNSAMPLED_GENES_EXPRESSION_LOADING = `${GENES}/downsampledGenesExpressionLoading`;

/**
 * Downsampled genes expression loaded
 */
const DOWNSAMPLED_GENES_EXPRESSION_LOADED = `${GENES}/downsampledGenesExpressionLoaded`;

/**
 * Downsampled genes expression loaded
 */
const DOWNSAMPLED_GENES_EXPRESSION_ERROR = `${GENES}/downsampledGenesExpressionError`;

export {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  DOWNSAMPLED_GENES_EXPRESSION_LOADING,
  DOWNSAMPLED_GENES_EXPRESSION_LOADED,
  DOWNSAMPLED_GENES_EXPRESSION_ERROR,
};
