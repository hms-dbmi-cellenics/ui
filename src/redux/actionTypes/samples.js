const SAMPLES = 'samples';

/**
 * Save sample.
 */
const SAMPLES_CREATE = `${SAMPLES}/create`;

/**
 * Update sample.
 */
const SAMPLES_UPDATE = `${SAMPLES}/update`;

/**
 * Delete sample.
 */
const SAMPLES_DELETE_API_V2 = `${SAMPLES}/deleteApiV2`;

/**
 * Delete sample following apiv1 version, remove when doing the final step of cleaning up v1 code.
 */
const SAMPLES_DELETE_API_V1 = `${SAMPLES}/deleteApiV1`;

/**
 * Update files in sample.
 */
const SAMPLES_FILE_UPDATE = `${SAMPLES}/fileUpdate`;

/**
 * Load sample.
 */
const SAMPLES_LOADED = `${SAMPLES}/loaded`;

/**
 * Saving samples.
 */
const SAMPLES_SAVING = `${SAMPLES}/saving`;

/**
 * Samples saved successfully.
 */
const SAMPLES_SAVED = `${SAMPLES}/saved`;

/**
 * Error loading/saving sample.
 */
const SAMPLES_ERROR = `${SAMPLES}/error`;

/**
 * Delete metada from sample.
 */
const SAMPLES_METADATA_DELETE = `${SAMPLES}/metadataDelete`;

const SAMPLES_LOADING = `${SAMPLES}/loading`;
export {
  SAMPLES_CREATE,
  SAMPLES_UPDATE,
  SAMPLES_DELETE_API_V2,
  SAMPLES_DELETE_API_V1,
  SAMPLES_FILE_UPDATE,
  SAMPLES_LOADED,
  SAMPLES_SAVING,
  SAMPLES_ERROR,
  SAMPLES_SAVED,
  SAMPLES_METADATA_DELETE,
  SAMPLES_LOADING,
};
