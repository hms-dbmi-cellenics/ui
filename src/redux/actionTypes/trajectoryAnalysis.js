const TRAJECTORY_ANALYSIS = 'trajectoryAnalysis';

/**
 * Loading trajectory analysis data
 */
const TRAJECTORY_ANALYSIS_LOADING = `${TRAJECTORY_ANALYSIS}/loading`;

/**
 * Trajectory analysis data loaded
 */
const TRAJECTORY_ANALYSIS_LOADED = `${TRAJECTORY_ANALYSIS}/loaded`;

/**
 * Error loading data
 */
const TRAJECTORY_ANALYSIS_ERROR = `${TRAJECTORY_ANALYSIS}/error`;

export {
  TRAJECTORY_ANALYSIS_LOADING,
  TRAJECTORY_ANALYSIS_LOADED,
  TRAJECTORY_ANALYSIS_ERROR,
};
