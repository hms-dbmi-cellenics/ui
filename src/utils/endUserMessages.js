export default {
  ACCOUNT_DETAILS_UPDATED: 'Account details updated successfully.',
  CONNECTION_ERROR: 'Could not connect to the server. Check your internet connection and refresh the page.',
  DELETING_PROJECT: 'Deleting project...',
  DELETING_SAMPLE: 'Deleting sample...',
  EMPTY_CLUSTER_NOT_CREATED: 'Cannot create a cluster that would be empty.',
  ERROR_DELETING_PROJECT: 'We couldn\'t delete your project.',
  ERROR_DELETING_SAMPLES: 'We couldn\'t delete samples from your project.',
  ERROR_DOWNLOADING_DATA: 'Error downloading data',
  ERROR_FETCHING_CELL_SETS: 'We couldn\'t get the list of cell sets.',
  ERROR_FETCHING_EXPERIMENTS: 'We couldn\'t get the list of experiments.',
  ERROR_FETCHING_PLOT_CONFIG: 'We couldn\'t get the plot settings.',
  ERROR_FETCHING_PLOT_DATA: 'We couldn\'t get the plot data.',
  ERROR_FETCHING_PROCESSING: 'We couldn\'t get the filter settings.',
  ERROR_FETCHING_PROJECTS: 'We couldn\'t get the list of projects.',
  ERROR_FETCHING_SAMPLES: 'We couln\'t get your samples.',
  ERROR_FETCHING_STATUS: 'We couldn\'t get the status of pending jobs.',
  ERROR_FETCHING_BACKEND_STATUS: 'We couldn\'t get the status of processing the experiment\'s processing run.',
  ERROR_NOT_SIGNED_IN: 'You need to sign in first. Click Sign in on the top right to begin.',
  ERROR_SAVING: 'We couldn\'t save your data.',
  ERROR_STARTING_PIPLELINE: 'We couldn\'t start working on your data.',
  ERROR_SIGN_IN: 'We couldn\'t sign you in. Please try again.',
  FEEDBACK_ERROR: 'We couldn\'t process your feedback. Please try submitting again.',
  FEEDBACK_SUCCESSFUL: 'Your feedback has been successfully sent. Thank you!',
  REFERRAL_ERROR: 'We couldn\'t process your referral. Please try submitting again.',
  REFERRAL_EMAIL_ERROR: 'The error you submitted was invalid. Please check again.',
  REFERRAL_SUCCESSFUL: 'Your invitation has been successfully sent. Thank you!',
  SUCCESS_NEW_CLUSTER_CREATED: 'Created a new cluster. Check the `Data Management` window.',
  SUCCESS_CELL_SETS_RECLUSTERED: 'Cell sets have been reclustered. Your cell sets have changed.',
  NO_CLUSTERS: 'No custom cell sets available, to create more go to data-exploration.',
  SAVING_SAMPLE: 'Saving sample...',
  SAVING_PROJECT: 'Saving project...',
};
