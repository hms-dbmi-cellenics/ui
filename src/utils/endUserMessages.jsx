export default {
  ACCOUNT_DETAILS_UPDATED: 'Account details updated successfully.',
  CONNECTION_ERROR: 'Could not connect to the server. Check your internet connection and refresh the page.',
  DELETING_PROJECT: 'Deleting project...',
  DELETING_SAMPLE: 'Deleting sample...',
  ERROR_CREATING_SAMPLE: 'We couldn\'t create your sample. Please try uploading it again',
  ERROR_CREATING_PROJECT: 'We couldn\'t create your project.',
  ERROR_LOADING_PROJECT: 'We couldn\'t load your project.',
  EMPTY_CLUSTER_NOT_CREATED: 'Cannot create a cluster that would be empty.',
  ERROR_DELETING_PROJECT: 'We couldn\'t delete your project.',
  ERROR_DELETING_SAMPLES: 'We couldn\'t delete samples from your project.',
  ERROR_DOWNLOADING_DATA: 'Error downloading data.',
  ERROR_FETCHING_CELL_SETS: 'We couldn\'t get the list of cell sets.',
  ERROR_FETCHING_EXPERIMENTS: 'We couldn\'t get the list of experiments.',
  ERROR_FETCHING_PLOT_CONFIG: 'We couldn\'t get the plot settings.',
  ERROR_FETCHING_PLOT_DATA: 'We couldn\'t get the plot data.',
  ERROR_FETCHING_NORMALIZED_EXPRESSION_MATRIX: 'We couldn\'t download the normalized expression matrix.',
  ERROR_NO_MATCHING_CELLS_NORMALIZED_EXPRESSION_MATRIX: 'No cells match requested filters.',
  ERROR_FETCHING_PROCESSING: 'We couldn\'t get the filter settings.',
  ERROR_FETCHING_DATA_PROCESSING_SETTINGS: 'We couldn\'t get the data processing settings.',
  ERROR_FETCHING_PROJECTS: 'We couldn\'t get the list of projects.',
  ERROR_FETCHING_SAMPLES: 'We couln\'t get your samples.',
  ERROR_FETCHING_STATUS: 'We couldn\'t get the status of pending jobs.',
  ERROR_FETCHING_BACKEND_STATUS: 'We couldn\'t get the status of the experiment\'s data processing.',
  ERROR_FETCH_BACKGROUND_GENE_EXP: 'We couldn\'t get the background gene expression',
  ERROR_FETCH_MARKER_GENES: 'We couldn\'t get the marker genes',
  ERROR_FILE_CATEGORY: 'The selected file name does not match the expected category.',
  ERROR_FILES_FOLDER: 'Only files contained in a folder are accepted.',
  ERROR_METADATA_MULTIPLE_FILES: 'You need to upload a single file in .tsv format',
  ERROR_METADATA_WRONG_FORMAT: 'The file does not seem to be in the correct tsv key-value format.',
  ERROR_CELL_LEVEL_COLUMN: 'The .tsv file needs to contain the column "barcode"',
  ERROR_OBJ2S_MULTIPLE_FILES: 'Only a single file is accepted for uploads.',
  ERROR_OBJ2S_EXISTING_FILE: 'Please remove existing file before uploading a new one.',
  ERROR_OBJ2S_MAX_FILE_SIZE: 'file size must be less than 15 GB.',
  ERROR_NOT_SIGNED_IN: 'You need to sign in first. Click Sign in on the top right to begin.',
  ERROR_NO_PERMISSIONS: 'You do not have permissions to perform that action.',
  ERROR_LAUNCH_PATHWAY: 'We couldn\'t launch pathway analysis',
  ERROR_SAVING: 'We couldn\'t save your data.',
  ERROR_BEGIN_SAMPLE_FILE_UPLOAD: 'We couldn\'t begin the files upload',
  ERROR_UPDATE_SERVER_ON_UPLOAD_STATE: 'We couldn\'t update the upload state in our servers. Try refreshing the page.',
  ERROR_SAMPLE_TECHNOLOGY: 'Selected technology does not match across all samples. Select the right technology and try again.',
  ERROR_SAVING_PLOT_CONFIG: 'We couldn\'t save your plot configuration.',
  ERROR_STARTING_PIPLELINE: 'We couldn\'t start working on your data.',
  ERROR_SIGN_IN: 'We couldn\'t sign you in. Please try again.',
  ERROR_DOWNLOADING_OBJ2S: 'We couldn\'t get the processed object.',
  FEEDBACK_ERROR: 'We couldn\'t process your feedback. Please try submitting again.',
  FEEDBACK_SUCCESSFUL: 'Your feedback has been successfully sent. Thank you!',
  REFERRAL_ERROR: 'We couldn\'t process your referral. Please try submitting again.',
  REFERRAL_EMAIL_ERROR: 'The error you submitted was invalid. Please check again.',
  REFERRAL_SUCCESSFUL: 'Your invitation has been successfully sent. Thank you!',
  SUCCESS_NEW_CLUSTER_CREATED: (
    <>
      Created a new cluster. Check &nbsp;
      <b>Custom cell sets</b>
      &nbsp; in the &nbsp;
      <b>Cell sets and Metadata </b>
      tile.
    </>),
  SUCCESS_CELL_SETS_RECLUSTERED: 'Cell sets have been reclustered. Your cell sets have changed.',
  SUCCESS_CELL_SETS_ANNOTATED: 'Cell sets have been annotated. Your cell sets have changed.',
  WORK_REQUEST_TIMED_OUT_RETRY: 'The task is taking longer than expected, please retry.',
  SAVING_SAMPLE: 'Saving sample...',
  SAVING_PROJECT: 'Saving project...',
  SHARE_SUCESS_NOTIFICATION_FAILURE: 'The project was shared, but we werenâ€™t able to notify the new collaborator',
  SHARE_FAILURE: 'We weren\'t able to share the project',
  ERROR_CLONING_DEFAULT: 'We couldn\'t clone the project.',
  ERROR_CLONING_PIPELINE_LOCKED: 'We couldn\'t clone the project because data processing is currently running, wait until it finishes and try again',
  ERROR_CELL_SETS_ANNOTATION_FAILED: 'ScType annotation unsuccessful. Please ensure dataset contains gene symbols.',
};
