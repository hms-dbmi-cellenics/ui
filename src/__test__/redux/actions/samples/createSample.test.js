import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { v4 as uuidv4 } from 'uuid';

import createSample from 'redux/actions/samples/createSample';
import initialSampleState from 'redux/reducers/samples/initialState';
import initialProjectState, { projectTemplate } from 'redux/reducers/projects/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';
import endUserMessages from 'utils/endUserMessages';

import { SAMPLES_CREATE, SAMPLES_SAVING, SAMPLES_ERROR } from 'redux/actionTypes/samples';

import pushNotificationMessage from 'utils/pushNotificationMessage';

pushNotificationMessage.mockImplementation(() => async () => { });

enableFetchMocks();

const mockStore = configureStore([thunk]);

jest.mock('uuid');
const sampleUuid = 'abc123';
uuidv4.mockImplementation(() => sampleUuid);

const sampleName = 'test sample';

describe('createSample action', () => {
  const projectUuid = 'qwe234';
  const experimentId = 'exp234';

  const mockType = '10x Chromium';

  const mockProject = {
    ...projectTemplate,
    name: 'test project',
    uuid: projectUuid,
    experiments: [experimentId],
  };

  const mockExperiment = {
    ...experimentTemplate,
    name: 'Experiment 1',
    id: experimentId,
  };

  const initialState = {
    samples: { ...initialSampleState },
    experiments: {
      ...initialExperimentState,
      [experimentId]: mockExperiment,
    },
    projects: {
      ...initialProjectState,
      ids: [projectUuid],
      [projectUuid]: mockProject,
    },
  };

  let store;

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());

    fetchMock.resetMocks();
    fetchMock.doMock();

    store = mockStore(initialState);
  });

  it('Runs correctly', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    const newUuid = await store.dispatch(createSample(projectUuid, sampleName, mockType));

    // Fetch call is made
    const fetchMockFirstCall = fetchMock.mock.calls[0];

    const { body: fetchBody, method: fetchMethod } = fetchMockFirstCall[1];
    expect(fetchMockFirstCall[0]).toEqual(`http://localhost:3000/v1/projects/${projectUuid}/${mockProject.experiments[0]}/samples`);

    expect(fetchMethod).toEqual('POST');
    expect(JSON.parse(fetchBody)).toMatchSnapshot();

    // Sends correct actions
    const actions = store.getActions();

    expect(actions[0].type).toEqual(SAMPLES_SAVING);
    expect(actions[1].type).toEqual(SAMPLES_CREATE);

    // Returns a new sampleUuid
    expect(newUuid).toEqual(sampleUuid);
  });

  it('Shows error message and throws an error when there is a fetch error', async () => {
    const fetchErrorMessage = 'someFetchError';

    fetchMock.mockResponse(JSON.stringify({ message: fetchErrorMessage }), { url: 'mockedUrl', status: 400 });

    let newUuid;

    // Fails with error message we sent in response to fetch
    await expect(async () => {
      newUuid = await store.dispatch(
        createSample(projectUuid, sampleName, mockType),
      );
    }).rejects.toThrowError(fetchErrorMessage);

    // Sends correct actions
    const actions = store.getActions();

    expect(actions[0].type).toEqual(SAMPLES_SAVING);
    expect(actions[1].type).toEqual(SAMPLES_ERROR);

    // Check no other action was sent
    expect(actions).toHaveLength(2);

    expect(pushNotificationMessage).toHaveBeenCalledWith('error', endUserMessages.ERROR_CREATING_SAMPLE);

    // It should not return a uuid
    expect(newUuid).toBeUndefined();
  });
});
