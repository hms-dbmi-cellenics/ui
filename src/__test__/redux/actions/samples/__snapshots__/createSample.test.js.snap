// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSample action Runs correctly 1`] = `
Object {
  "complete": false,
  "createdDate": "2020-01-01T00:00:00.000Z",
  "error": false,
  "fileNames": Array [],
  "files": Object {},
  "lastModified": "2020-01-01T00:00:00.000Z",
  "metadata": Object {},
  "name": "test sample",
  "projectUuid": "qwe234",
  "type": "10X Chromium",
  "uuid": "abc123",
}
`;

exports[`createSample action Works correctly for api v2 1`] = `
Object {
  "name": "test sample",
  "sampleTechnology": "10x",
}
`;
